consensus/src/metrics_safety_rules.rs

```diff
@@ -125,3 +142,111 @@ impl TSafetyRules for MetricsSafetyRules {
+    use diem_types::{
+        block_info::BlockInfo,
+        epoch_change::EpochChangeProof,
+        ledger_info::{LedgerInfo, LedgerInfoWithSignatures},
+        waypoint::Waypoint,
+    };
+    use safety_rules::{ConsensusState, Error, TSafetyRules};
+
+    pub struct MockSafetyRules {
+        // number of initialize() calls
+        init_calls: i32,
+
+        // max initialize() calls to complete perform_initialize()
+        max_init_calls: i32,
+    }
+
+    impl MockSafetyRules {
+        pub fn new(init_calls: i32, max_init_calls: i32) -> Self {
+            Self {
+                init_calls,
+                max_init_calls,
+            }
+        }
+    }
+
+    impl TSafetyRules for MockSafetyRules {
+        fn consensus_state(&mut self) -> Result<ConsensusState, Error> {
+            let safety_data = SafetyData::new(0, 0, 0, 0, None);
+            let waypoint = Waypoint::new_any(&LedgerInfo::new(
+                BlockInfo::new(0, 0, HashValue::zero(), HashValue::zero(), 0, 0, None),
+                HashValue::zero(),
+            ));
+            Ok(ConsensusState::new(safety_data, waypoint, true))
+        }
+
+        fn initialize(&mut self, _: &EpochChangeProof) -> Result<(), Error> {
+            self.init_calls += 1;
+            if self.init_calls < self.max_init_calls {
+                return Err(Error::WaypointOutOfDate(
+                    (self.init_calls - 1) as u64,
+                    self.init_calls as u64,
+                    self.max_init_calls as u64,
+                    self.init_calls as u64,
+                ));
+            }
+            Ok(())
+        }
+
+        fn construct_and_sign_vote(&mut self, _: &MaybeSignedVoteProposal) -> Result<Vote, Error> {
+            unimplemented!()
+        }
+
+        fn sign_proposal(&mut self, _: &BlockData) -> Result<Ed25519Signature, Error> {
+            unimplemented!()
+        }
+
+        fn sign_timeout(&mut self, _: &Timeout) -> Result<Ed25519Signature, Error> {
+            unimplemented!()
+        }
+
+        fn sign_timeout_with_qc(
+            &mut self,
+            _: &TwoChainTimeout,
+            _: Option<&TwoChainTimeoutCertificate>,
+        ) -> Result<Ed25519Signature, Error> {
+            unimplemented!()
+        }
+
+        fn construct_and_sign_vote_two_chain(
+            &mut self,
+            _: &MaybeSignedVoteProposal,
+            _: Option<&TwoChainTimeoutCertificate>,
+        ) -> Result<Vote, Error> {
+            unimplemented!()
+        }
+
+        fn sign_commit_vote(
+            &mut self,
+            _: LedgerInfoWithSignatures,
+            _: LedgerInfo,
+        ) -> Result<Ed25519Signature, Error> {
+            unimplemented!()
+        }
+    }
+
+    #[test]
+    fn test_perform_initialize() {
+        let (_, mock_storage) = EmptyStorage::start_for_testing();
+        let mock_safety_rules =
+            MockSafetyRules::new(0 /* init_calls */, 10 /* max_init_calls */);
```

Comment:
btw I think those comments are unnecessary given IDE can prompt the type information inline

Reply:
Removed unnecessary comments.